---
title: "from otu to similarity matrix by blast"
author: "xyz"
date: "2021/4/6"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# read Enterobacteriaceae relatives

```{r}
library(Biostrings)
En <- readRDS("../temp/En.rds")
EnSeq <- readRDS("../temp/EnSeq.rds")
```

# read E. coli sequence

```{r}
ecoliEx <-
  readDNAStringSet("../temp/Escherichia-ShigellASV.fasta")
ecoliExWithTax <-
  read.csv("../table/Escherichia-ShigellASVwithTax.csv")
names(ecoliEx) <- ecoliExWithTax$ASV.IX
ecoliStandard <-
  readDNAStringSet("../temp/ecoli16sSeq515F907R370.fasta")

# the sequence of E_coli_W is the same as ASV2 in ecoliEx
ecoliStandard[[1]] == ecoliEx[[1]]

allEn <- c(ecoliEx, EnSeq)
writeXStringSet(allEn, "../temp/allEn.fasta")

annotationDf <- data.frame(
  Taxa = names(allEn),
  Source = c(rep("E. coli", length(ecoliEx)),
             str_sub(En$Genus, 4)),
  Species = c(names(ecoliEx),
              str_sub(En$Species, 4)),
  Count = c(ecoliExWithTax$totalCount, rep(1, length(EnSeq)))
)
write.csv(annotationDf, "../temp/allEn.Annotation.csv", row.names = F)
```

# get Nucleotide Similarity by blast

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2

# one sequence may match another more than 1 time
blastn -query ../temp/allEn.fasta -subject ../temp/allEn.fasta -outfmt 6  \
  -task blastn \
  -max_hsps 1000 \
  -max_target_seqs 1000 \
  -evalue 1000 -perc_identity 0 \
  -out "../temp/allEn2allEn.txt"
```

```{r}
library(tidyverse)
blastColnames <-
  c(
    "qseqid",
    "sseqid",
    "pident",
    "length",
    "mismatch",
    "gapopen",
    "qstart",
    "qend",
    "sstart",
    "send",
    "evalue",
    "bitscore"
  )
df <- read.delim("../temp/allEn2allEn.txt", header = F)
colnames(df) <- blastColnames

# arrange by match length
df2 <- arrange(df, qseqid, sseqid, desc(length))
# reduplicate by keep the longest match length
df2 <- df2[!duplicated(df2[, 1:2]), ]
write.csv(df2,
          "../table/All Enterobacteriaceae Nucleotide Similarity by blast.csv",
          row.names = F)

# the alignment of E.coli Sequence
df3 <-
  df2[str_detect(df2$qseqid, "ASV") & str_detect(df2$sseqid, "ASV"), ]
nrow(df3) == 19 * 19

# convert to matrix format
df4 <- pivot_wider(df3[, 1:3], names_from = sseqid, values_from = pident)
colnames(df4)[1] <- ""
write.csv(df4,
          "../table/E coli Nucleotide Similarity Matrix by blast.csv",
          row.names = F)
```
