---
title: "Qiime2 workflow"
author: "xyz"
date: "2021/3/31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# make meta data

```{r}
library(stringr)
fileName<-dir("../data/fastq/")
# 454Reads.MID_43.sff.fq is empty
fileName<-fileName[fileName!="454Reads.MID_43.sff.fq"]
sampleID<-str_split(fileName,"\\.",simplify = T)[,2]
filePath<-paste0("/mnt/e/xiongyi/454Data_JiaLab-2020.12.31/from454sff/data/fastq/",fileName)
df <- data.frame(
  `sample-id` = sampleID,
  `absolute-filepath` = filePath,
  stringsAsFactors = F
)
colnames(df) <- c("sample-id", "absolute-filepath")
write.table(
  df,
  "../data/metadata.tsv",
  quote = F,
  row.names = F,
  sep = "\t"
)
```

# import to qiime

```{bash import}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime tools import \
  --type 'SampleData[SequencesWithQuality]' \
  --input-path ../data/metadata.tsv \
  --output-path ../data/fastq.qza \
  --input-format SingleEndFastqManifestPhred33V2

# check quality
qiime demux summarize \
    --i-data ../data/fastq.qza \
    --o-visualization ../temp/qcSummary.qzv
```

# remove primer

515F GTGCCAGCMGCCGCGG
907R CCGTCAATTCMTTTRAGTTT

[linked-adapters](https://cutadapt.readthedocs.io/en/stable/guide.html#linked-adapters)

remove sequence shorter than 300

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime cutadapt trim-single \
  --p-cores 6 \
  --i-demultiplexed-sequences ../data/fastq.qza \
  --p-adapter GTGCCAGCMGCCGCGG...AAACTYAAAKGAATTGACGG \
  --p-error-rate 0.2 \
  --p-minimum-length 300 \
  --o-trimmed-sequences ../temp/fastq.primer_trimed.qza \
  --verbose &> ../temp/fastq.primer_trimed.log
  
qiime demux summarize \
    --i-data ../temp/fastq.primer_trimed.qza \
    --o-visualization ../temp/fastq.primer_trimed.qcSummary.qzv
```

# denoise by DaDa2

[DADA2 denoise](https://mp.weixin.qq.com/s/s_lU38X56zLWKFMHmHUzCg)

## no truncate

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime dada2 denoise-pyro \
  --p-trunc-len 0 \
  --i-demultiplexed-seqs ../temp/fastq.primer_trimed.qza \
  --p-n-threads 6 \
  --o-table ../temp/fastq.denoised.qza \
  --o-representative-sequences ../temp/fastq.rep-seqs.qza \
  --o-denoising-stats ../temp/fastq.denoising-stats.qza

qiime metadata tabulate \
    --m-input-file ../temp/fastq.denoising-stats.qza \
    --o-visualization ../temp/fastq.denoising-stats.qzv
```

## truncate to 370

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime dada2 denoise-pyro \
  --p-trunc-len 370 \
  --i-demultiplexed-seqs ../temp/fastq.primer_trimed.qza \
  --p-n-threads 6 \
  --o-table ../temp/fastq370.table.qza \
  --o-representative-sequences ../temp/fastq370.rep-seqs.qza \
  --o-denoising-stats ../temp/fastq370.denoising-stats.qza

qiime metadata tabulate \
    --m-input-file ../temp/fastq370.denoising-stats.qza \
    --o-visualization ../temp/fastq370.denoising-stats.qzv
```

# Cluster Feature 

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime vsearch cluster-features-de-novo \
  --i-sequences ../temp/fastq370.rep-seqs.qza \
  --i-table ../temp/fastq370.table.qza \
  --p-perc-identity 0.97 \
  --o-clustered-table ../temp/fastq370.97.table.qza \
  --o-clustered-sequences ../temp/fastq370.97.seq.qza \
  --p-threads 6
```

# species annotation

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime feature-classifier classify-sklearn \
  --i-reads ../temp/fastq370.97.seq.qza \
  --i-classifier ../../2021xiongyiAnalysis/db/SILVA_138/silva138_515F_907R_classifier.qza \
  --o-classification ../temp/denovo.clustered.naive_bayes_taxonomy.qza \
  --p-n-jobs 1 \
  &
qiime feature-classifier classify-consensus-blast \
  --i-query ../temp/fastq370.97.seq.qza \
  --i-reference-reads ../../2021xiongyiAnalysis/db/SILVA_138/silva-138-99-seqs.qza \
  --i-reference-taxonomy ../../2021xiongyiAnalysis/db/SILVA_138/silva-138-99-tax.qza \
  --o-classification ../temp/denovo.clustered.blast_taxonomy.qza
```

# export OTU table and sequence

```{bash}
source ~/miniconda3/etc/profile.d/conda.sh
conda activate qiime2
qiime tools export \
  --input-path ../temp/fastq370.97.seq.qza \
  --output-path ../temp

qiime tools export \
  --input-path ../temp/fastq370.97.table.qza \
  --output-path ../temp
  
biom convert -i ../temp/feature-table.biom \
  -o ../temp/feature-table.tsv --to-tsv
  
qiime tools export \
  --input-path ../temp/denovo.clustered.naive_bayes_taxonomy.qza \
  --output-path ../temp/
```

# Select OTU with more than 1 sequence

```{r}
library(tidyverse)
otu <- read.table("../temp/feature-table.tsv",header = T,
                  sep = "\t",comment.char="",skip=1)
colnames(otu)[1]<-"OTU.ID"
tax <- read.table("../temp/taxonomy.tsv",
                  header = T,
                  sep = "\t")
# 1 OTU without annotation
sum(tax$Taxon == "Unassigned")
# tax<-tax[tax$Taxon != "Unassigned",]
sum(is.na(tax$Taxon))
rankTaxon <- data.frame(Taxon = tax$Taxon)
rankTaxon <-
  separate(rankTaxon, Taxon, sep = ";", into = as.character(1:7))
colnames(rankTaxon) <-
  c("Kingdom",
    "Phylum",
    "Class",
    "Order",
    "Family",
    "Genus",
    "Species")
rankTaxon <- cbind(id = tax$Feature.ID,
                   rankTaxon,
                   Confidence = tax$Confidence)
otuWithTax <- right_join(otu, rankTaxon, by = c("OTU.ID" = "id"))
totalCount <- rowSums(otuWithTax[, 2:ncol(otu)])
totalPercent <- totalCount / sum(totalCount) * 100
otuWithTax$totalCount <- totalCount
otuWithTax$totalPercent <- totalPercent
otuWithTax <- arrange(otuWithTax, desc(totalCount))
otuWithTax$OTU.IX<-paste0("OTU",1:nrow(otuWithTax))
otuWithTax<-otuWithTax[otuWithTax$totalCount > 1,]
write.csv(otuWithTax,
          "../table/OTUwithTax97ByNaive_bayesSortByOneMore.csv",
          row.names = F)
# 6524 OTU with more than 1 sequences 
sum(otuWithTax$totalCount > 1)
```